Glimmer component

wycats opened the pull request:
Don't merge this yet, until the following items are complete:
- [x] reintroduce mergeAttrs & normalization for identity elements (`<my-foo>` top level inside `<my-foo>`) - done in d280c5353c3031272a3fda59e619e1d8f6fc6e02
- [x] remove attributes hook (post-processing for top-level elements) - done in d280c5353c3031272a3fda59e619e1d8f6fc6e02
- [x] figure out why `component.element` may not be set on `GlimmerComponent` fa01e2424d1e699c43abfaed37266339936c3305
- [x] identify and fix scope bug in identity elements c745c4766c371a3495049a9769876a6680ddbe85
- [x] make regular elements like identity elements -- done in ca436a2b1d4d488a6aeadd9ba7e6ab170e3a5e49 and 87e021a83559001dec692e6872a69ce2752f8f59
- [x] align fallback semantics with htmlbars (`<input value={{bar}} />`)
- [x] Re-enable previously skipped tests related to angle bracket components.
- [x] Fix and incorporate currently failing test from https://github.com/emberjs/ember.js/pull/11991 (test added in 6f359d2, fixed previously in c745c47)
- [x] Make sure `{{#with}}<div {{action}}></div>{{/with}}` works (64247979b8d112de7a89c259ca3e6edc9061abca)
- [x] Error when you use a fragment inside `GlimmerComponent` (5b686d44e3cadb174ed30a2f00e605eca1ebf37f)
  - [x] Better error when fragment comes from `{{action "foo"}}` or `style={{{bar}}}`
- [x] Make sure existing uses of already-registered web components continue to work (974493f0ddef29f607498d385d3ff3258ba00894)
- [x] all green??????????

Post-merge:
- [ ] Fix top-level `{{partial}}` case (0604db16ea85496b294d31f4289fbf1687924e89)
- [ ] Make sure that people don't accidentally use old-style APIs for top-level elements in GlimmerComponent (`tagName`)
- [ ] Refactor to separate conceptually different code paths that happen to share code, such as...
  - [ ] build-component-template
  - [ ] component node manager?
- [ ] make recursive invocation work (`<my-foo>` component as the top-level element inside `<my-bar>`'s layout)
- [ ] discuss a path for properties vs. attributes being plausibly aligned between normal elements and components

tomdale committed:
Remove unused view mixin

tomdale committed:
Move Ember.Component to components/ directory

tomdale committed:
wip

wycats committed:
Create GlimmerComponent subclass

Commit:
Fix tests

1. Make sure tests that test Glimmer Components are feature flagged.
2. Update a test to use `init` instead of `didInitAttr` due to a change to
    the lifecycle that landed on master.

Commit:
Fix attr setting in Glimmer Components

Previously, Glimmer components got their attrs set after positional
attrs were applied. The `didInitAttrs` hook was used to indicate that
the full set of attributes were applied.

Recently, positional attrs were changed to ensure that the full set of
attrs are always applied before `init`, eliminating the need for
`didInitAttrs`.

However, the existing code that applied attrs on init only did so for
non-angle bracket components (to avoid double-setting on new-world
features). Consequently, angle bracket components *never* got their
attrs set.

This commit ensures that `attrs` are set on all components before init,
regardless of style.

Commit:
Show error for component class/invocation mismatch

1. {{my-component ...}} curly braces can only be used on legacy Component
2. <my-component ...> angle brackets can only be used on GlimmerComponent

TBD: rule number 2 might be loosened with an explicit opt-in later,
depending on how painful this transition turns out to be for large apps.

It might be better to provide automated rewriting through ember-watson,
which would do the job quickly, once and for all, rather than leaving
large apps in a semantic limbo with one foot in the old world and
another foot in the new world.

Commit:
Re-enable skipped tests for GlimmerComponents

They were disabled due to a bug in the feature flag infrastructure; but
it has since been fixed and everything appears to be working again here.

funtusov committed:
Failing test for computed property alias in angle-bracket component

Commit:
Remove incorrect attributes

This code was previously telling HTMLBars to post-process content
templates to add top-level attributes. That is not correct (it only
makes sense to post-process layouts).

Commit:
This commit fixes a top-level element bug

If the layout for `<my-component>` has `<my-component>` as its root
element, this is a special case. The semantics of this case are intended
to be the same as a top-level `<div>`.

However, in an attempt to share code, we inadvertantly introduced a bug
where we installed the dynamic attributes defined by JS APIs twice (via
`normalizeComponentAttributes` and post-processing of the parent
template).

The fix, therefore, is to not do anything special with dynamic
attributes and rely on the same post-processing step as the `<div>`
case.

Commit:
Add some new (failing) tests for component bugs

stefanpenner commented on packages/ember-htmlbars/lib/hooks/bind-self.js:
```diff
@@ -19,7 +19,13 @@ export default function bindSelf(env, scope, _self) {
     newStream(scope.locals, 'controller', scope.locals.view.getKey('controller'));
-    newStream(scope, 'self', scope.locals.view.getKey('context'), null, true);
+
+    if (self.isGlimmerComponent) {
```
should we instead use polymorphism here? Something like `this._setupSelfStream()` and let the subclasses refine further?

stefanpenner commented on packages/ember-htmlbars/lib/node-managers/component-node-manager.js:
```diff
@@ -301,13 +310,13 @@ ComponentNodeManager.prototype.destroy = function() {
 export function createComponent(_component, isAngleBracket, _props, renderNode, env, attrs = {}, proto = _component.proto()) {
   let props = assign({}, _props);
 
+  let snapshot = takeSnapshot(attrs);
```
curious why this was pulled out of the branch

wycats replied:
8dc7d58cd4c0f45cc5aaf79c3ccf03d10dd7445f

stefanpenner replied:
ah thanks...

(would be nice if GH should which commit a specific change in a diff came from)

runspired commented:
:D

Commit:
Remove post-processing ("attributes") hook

This was previously used for attaching outer attributes as well as
system attributes (e.g. class="ember-view" id="ember123") onto regular
HTML elements at the root, once they have been rendered.

This results in multiple AttrNodes for a single attribute, which means
that there is no guarantee that any particular AttrNode will actually
"win". This incorrect strategy "works" a surprising amount of the time,
but the failure modes are unacceptable. (See new tests introduced in the
previous commit.)

This is in anticipation of making all top-level elements in a
component's layout dynamic, so that they can share attribute merging
logic with the "identity element" case (top-level `<my-component>` in
the layout for `my-component`).

Commit:
Fix GlimmerComponent not having `this.element` set

Commit:
Refactor test cases to not test exact output

This was annoying because the ordering of non-user attributes are
unimportant, but can change (and has changed) depending on the
implementation.

Commit:
Skip failing tests for now

Also re-enabling a test that was actually fixed in the previous commit.

Commit:
Fix GlimmerComponent scoping issue

See "non-block ... should have correct scope" test cases

chancancode commented:
#11728

Commit:
WIP: make regular elements like identity elements

Commit:
Merge remote-tracking branch 'origin/master' into glimmer-component

rwjblue commented:
We should address the plan (or intentional avoidance) of the nested component case (components in sub-dirs) https://github.com/emberjs/ember.js/issues/10456.

wycats commented:
@rwjblue seems legit. I assume you mean via documentation, right?

rwjblue commented:
@wycats - Yeah, we should have docs and tests that say exactly what we do and don't support subdirectory wise.  I'm not suggesting that we have to support subdirs for landing + enabling, but I'd like to be very clear about what we intend (and that we didn't "forget").

Commit:
Skip AST transform without Glimmer Components

To reduce the possibility of regressions when the
`ember-htmlbars-component-generation` flag is off, this commit moves an
important AST transformation behind the flag.

This AST transformation makes top-level elements in *all* templates
dynamic, and requires runtime work to restore the original semantics.

Moving the AST transformation behind the flag limits the impact of any
bugs in the runtime work.

Commit:
Make regular elements like identity elements

In Ember 1.x components, the top-level element was configured in
JavaScript using APIs like `classNames`, `attributeBindings`, `tagName`
etc.

Glimmer components move that configuration into a top-level element in
the template. However, the template compiler does not know how the
template will actually be used, so it makes all top-level elements
dynamic, and leaves it up to the runtime to figure out what to do.

If the template is being invoked for a curly component, the runtime
restores the original semantics, treating it just like a regular
element.

If the template is being invoked for an angle-bracket component, the
top-level element is treated as the component’s element.

This commit makes top-level `<div>`s work correctly in Glimmer
components (they become the component’s element), and work correctly
in curly components (they behave as before).

Both modifiers (`<div {{action “foo”}}>`) and triple-curly attributes
are not supported in top-level elements of Glimmer components. As a
result, if the AST transformation sees either of those two features,
it assumes the template is for a curly component, and does not make the
root element dynamic.

At the moment, it would be impossible to support either of those two
features, since the component AST node in HTMLBars do not support them.
Ultimately, we may want to support triple-curly attributes, but probably
not modifiers.

Note that this commit requires an update to HTMLBars master.

Commit:
Merge remote-tracking branch 'origin/master' into glimmer-component

Commit:
Merge pull request #11991 from funtusov/failing_angle_bracket_component_computed_function

Failing test for computed property alias in angle-bracket component

Commit:
Optimize the AST transform for top-level elements

Don’t bother to convert top-level elements into dynamic calls if they
aren’t the top-level element *of a template*.

Consider a template like this:

```
{{#each posts as |post|}}
<div>hello</div>
{{/each}}
```

The previous version of the transform would convert the nested template
(the <div>) into a dynamic call, because the AST transformation didn’t
differentiate between top-level templates and nested templates. It
didn’t have any other effects (the dynamic call is required to restore
the semantics of the original static form) but it unnecessarily bloats
templates and makes execution slower.

Commit:
Disallow fragments in GlimmerComponent for now

There are multiple competing proposals for how to represent fragments in
GlimmerComponents, and we have not yet settled on the exact solution.

In the interim, GlimmerComponents require a top-level wrapper element.
This commit also provides a good error message describing exactly what
the user did wrong. In the future, when fragments are allowed, we can
update the error message to suggest opting into a fragment.

Commit:
Implement nextElementSibling in JS

This feature works in all browsers in our support matrix except for
PhantomJS, so the Travis tests are failing.

Commit:
Merge remote-tracking branch 'origin/master' into glimmer-component

Commit:
Reorganize for transformation bug

The current ember-debug transformations do not support nested asserts
inside runInDebug.

Commit:
Make non-Ember-component <foo-bar> work

As of the previous commit, any use of `<foo-bar>` was assumed to be an
Ember Glimmer Component. This breaks code that was using dasherized
names either to refer to web components or just to produce regular HTML
elements. We believe that both of these are done in practice.

This commit allows `<foo-bar>` to be treated the same as `<div>` in all
currently implemented contexts.

Commit:
Test all styles of top-level element in <GC>

Previously, we had hardcoded tests for two styles of top-level element
in a GlimmerComponent. For the layout of `<foo-bar>`, we had tests for
the top-level element being `<foo-bar>` itself and being a `<div>`.

A recent commit allowed the top-level element to also be any dasherized
element that was not registered as an Ember component (aka “web
components”). This commit modifies the tests to enumerate all three
styles and runs the same set of tests for each.

chancancode commented:
We also need to double check the GlimmerComponent class is exported correctly (right module path and globally mode too?)

chadhietala commented:
Probably needs an RFC, but it would be great if glimmer components could optionally specified required types of attrs during dev.  It would make integrating components from addons a much more pleasant experience, might also have perf wins in prod due to the fact that the code was designed to be called consistently with the same object types. This would be like `propTypes` in react.

Commit:
Add failing test for top-level {{partial}}

Currently a top-level partial that happens to have a single top-level
element confuses the heuristics in the component hook.

Commit:
Merge remote-tracking branch 'origin/master' into glimmer-component

runspired commented:
@chadhietala +1 on that, I try to write helpful asserts but also hate writing them all the time.

rwjblue committed:
Merge branch 'master' into glimmer-component

rwjblue committed:
Do not export Ember.GlimmerComponent unless feature is enabled.
